---
title: "The Endometriosis-associated Microbiome"
name: "Samara Khan"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{The Endometriosis-associated Microbiome}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# Analysis objectives

1. Import, recode, and subset data from bugsigdb.org    
2. Create a table of studies
3. Calculate the frequency of appearance of each taxa in independent signatures and identify the most frequently reported taxa
4. Estimate the probability of the most frequently identified taxa occuring by chance

# Making sure packages are installed

```{r, eval=FALSE, messages=FALSE}
install.packages(c("devtools", "tidyverse", "kableExtra"))
devtools::install_github("waldronlab/bugSigSimple")
devtools::install_github("waldronlab/BugSigDBStats")
devtools::install_github("waldronlab/bugsigdbr")
```

# Load and subset data

```{r, messages=FALSE}
library(bugSigSimple)
library(BugSigDBStats)
library(bugsigdbr)
library(tidyverse)
library(stringr)
library(kableExtra)
library(dplyr)
```

```{r}
library(bugSigSimple)
dat <- bugsigdbr::importBugSigDB(cache= TRUE)
dim(dat)
names(dat)
```

# Subsetting

Subsetting only studies done on humans - 6 of 17 studies were excluded because they were either done on mice or not statistically reliable (Cregger et. al)

```{r, messages=FALSE}
subset.dat <-
  dat[which(
    dat$PMID == "30778155" |
      dat$PMID == "32192080" |
      dat$PMID == "31087436" |
      dat$PMID == "26901400" |
      dat$PMID == "33839907" |
      dat$PMID == "32046455" |
      dat$PMID == "33925708" |
      dat$PMID == "32299442" |
      dat$PMID == "33313185" |
      dat$PMID == "34268384" | dat$PMID == "33660232"
  ), ]
```

# All studies

Summary of studies and most frequent taxa increased and decreased in endometriosis patients for all studies

```{r, messages=FALSE}
bugSigSimple::createStudyTable(subset.dat)
getMostFrequentTaxa(subset.dat, n=30)
getMostFrequentTaxa(subset.dat, direction="UP")
getMostFrequentTaxa(subset.dat, direction="DOWN")
```

# Excluding feces samples

Summary of studies and most frequent taxa in only samples from female reproductive tract, excluding feces samples

```{r, messages=FALSE}
subset.dat2 <-
  dat[which(
    dat$PMID == "30778155" |
      dat$PMID == "32192080" |
      dat$PMID == "31087436" |
      dat$PMID == "26901400" |
      dat$PMID == "32046455" |
      dat$PMID == "33925708" |
      dat$PMID == "32299442" |
      dat$PMID == "33313185" | dat$PMID == "34268384"
  ), ]

reproductive_sigs <-
  subset.dat2[which(subset.dat2$`Body site` != "feces" |
                      is.na(subset.dat2$`Body site`)), ]

bugSigSimple::createStudyTable(reproductive_sigs)

allfreqs <- bugSigSimple::createTaxonTable(reproductive_sigs, n = 20) %>% #could change number
  arrange(I(decreased_signatures - increased_signatures))
incfreqs <- filter(allfreqs, I(increased_signatures - decreased_signatures) > 0)
decfreqs <- filter(allfreqs, I(increased_signatures - decreased_signatures) < 0)
kableExtra::kbl(allfreqs) %>%
  kable_paper("hover", full_width = FALSE)
```

## Load HMP data from curatedMetagenomicData

```{r, message=FALSE}
suppressPackageStartupMessages(library(curatedMetagenomicData))
# hmp_vagina_metadata <- filter(sampleMetadata, study_name == "HMP_2012" & body_site == "vagina")
all_healthy_adult_vagina_metadata <- filter(sampleMetadata, body_site == "vagina" & disease == "healthy" & age_category == "adult")
se <-
  curatedMetagenomicData::returnSamples(all_healthy_adult_vagina_metadata, dataType = "relative_abundance", counts = FALSE)
```

Get matrices of species and genus relative abundance
```{r}
allranks <- mia::splitByRanks(se)
#species_relab <- t(assay(allranks[["species"]]))
genus_relab <- t(assay(allranks[["genus"]]))
#family_relab <- t(assay(allranks[["family"]]))
```

Centered log-ratio transformation on genus level relative abundance:
```{r}
genus_relab_clr <- Hotelling::clr(genus_relab + 1)
```

```{r}
df <- dplyr::select(data.frame(genus_relab_clr), c("Streptococcus", "Lactobacillus"))
df$age <- se$age
df$BMI <- se$BMI
fit <- lm(Streptococcus ~ Lactobacillus + age + BMI, data = df)  
summary(fit)
summary(lm(Streptococcus ~ Lactobacillus, data = df))
wilcox.test(df$Streptococcus, df$Lactobacillus)
t.test(df$Streptococcus, df$Lactobacillus)
```
```{r}
secount <-
  curatedMetagenomicData::returnSamples(all_healthy_adult_vagina_metadata, dataType = "relative_abundance", counts = TRUE)
allranks <- mia::splitByRanks(secount)
genus_relab_count <- t(assay(allranks[["genus"]]))
df <- dplyr::select(data.frame(genus_relab_count), c("Streptococcus", "Lactobacillus"))
df$age <- secount$age
df$BMI <- secount$BMI
```

```{r}
library(pscl)
fit.ZInegbin <- pscl::zeroinfl(Streptococcus ~ Lactobacillus + age + BMI | 1,
                data = df,
                dist = "negbin")
```


# Create a data frame for regression analysis from healthy controls for log-linear model

```{r}
library(curatedMetagenomicData)
library(dplyr)
library(DT)
library(purrr)

taxon_strep <- c("1884" = "Streptococcus")
taxon_lact <- c("1578" = "Lactobacillus")

target_sample <- "vagina"
sample_metadata <- sampleMetadata %>% 
    filter(
        grepl(target_sample, body_site) | grepl(target_sample, body_subsite)
    ) %>% 
    discard(~all(is.na(.x))) %>% 
    as_tibble()
dim(sample_metadata)
unique(sample_metadata$study_name)

tse <- returnSamples(
    sampleMetadata = sample_metadata,
    dataType = "relative_abundance",
    counts = FALSE,
    rownames = "short"
)
tse
```


## Comparing mean to variance of the outcome variable

```{r}
library(ggplot2)
mean(df$Streptococcus)
var(df$Streptococcus)
fraczeroes <- nrow(df[df$Streptococcus == 0,])/nrow(df)
fraczeroes
```

#Fit log-linear models
```{r}
fit.pois <- glm(Streptococcus ~ Lactobacillus + age + BMI,
                data = df,
                family = poisson(link = "log"))

summary (fit.pois)

library(pscl)
fit.ZIpois <- zeroinfl(Streptococcus ~ Lactobacillus + age + BMI | 1,
                data = df,
                dist = "poisson")

summary (fit.ZIpois)

library(MASS)
fit.negbin <-
  glm.nb(
    Streptococcus ~ Lactobacillus + age + BMI,
    data = df,
    control = glm.control(
      epsilon = 1e-8,
      maxit = 50,
      trace = FALSE
    )
  )
library(broom)
tidy(fit.negbin)

fit.ZInegbin <- zeroinfl(Streptococcus ~ Lactobacillus + age + BMI | 1,
                data = df,
                dist = "negbin")

summary(fit.ZInegbin)
```



